From a8ac7e35dceda4dd732f2bff05bac51ffab5927f Mon Sep 17 00:00:00 2001
Subject: [PATCH] Fixing handling of github address

See bug #1019696
devscripts: uscan does not work with github anymore -
https://bugs.debian.org/1019696
---
 lib/Devscripts/Uscan/WatchLine.pm | 16 ++++++++++++++++
 lib/Devscripts/Uscan/http.pm      |  2 +-
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/lib/Devscripts/Uscan/WatchLine.pm b/lib/Devscripts/Uscan/WatchLine.pm
index 272a8d8..d5494f8 100644
--- a/lib/Devscripts/Uscan/WatchLine.pm
+++ b/lib/Devscripts/Uscan/WatchLine.pm
@@ -726,6 +726,22 @@ EOF
             $filepattern .= '(?:\?.*)?';
         }
 
+        # Handle github.com addresses specially
+        if (!$self->shared->{bare} and $self->mode ne 'git' and $base =~ m%^https?://github\.com/% and $base !~ m%tags$%) {
+            uscan_verbose "GitHub re-write rule.  Abandon all hope ye who enter here.";
+            my ($user, $repo) = $base =~ m%^https?://github\.com/([^\/]+)/([^\/]+)/%;
+            $base =~ s%^https?://github\.com/%https://api.github.com/repos/%;
+            if ($base =~ m%tags$%) {
+                # Yes, the full path is needed.  We can't use the provided filepattern as it needs to lack an extension.
+                $filepattern = "https://api\\.github\\.com/repos/$user/$repo/tarball/refs/tags/(?:[-_v]?(\\d[-+.:~\\da-zA-Z]*))";
+            } else {
+                $filepattern =~ s%^.*/\)?%%;
+                $filepattern =~ s%^%(?:https://api\\.github\\.com/repos/$user/$repo/tarball/(?:[-_v]?(\\d[-+.:~\\da-zA-Z]*))|https://github\\.com/$user/$repo/releases/download/v?(?:\\d[\\d\\.]*)/%;
+                $filepattern =~ s%$%)%;
+            }
+            $self->searchmode('plain');
+        }
+
         # Handle pypi.python.org addresses specially
         if (   !$self->shared->{bare}
             and $base =~ m%^https?://pypi\.python\.org/packages/source/%) {
diff --git a/lib/Devscripts/Uscan/http.pm b/lib/Devscripts/Uscan/http.pm
index 0da9798..c08bff3 100644
--- a/lib/Devscripts/Uscan/http.pm
+++ b/lib/Devscripts/Uscan/http.pm
@@ -482,7 +482,7 @@ sub parse_href {
             # exception, otherwise $mangled_version = 1
             $mangled_version = '';
         } else {
-            $mangled_version = join(".",
+            $mangled_version = join("",
                 map { $_ if defined($_) }
                   ref $match eq 'ARRAY' ? @$match : $href =~ m&^$_pattern$&);
         }
-- 
2.35.1

